// @progress/kendo-vue-charts/dist/es/components/value-axis-item/NotesLabel.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../base/ConfigurationComponent';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../utils/main';
/**
 * @hidden
 */
var ChartValueAxisNotesLabelVue2 = {
  name: 'KendoChartValueAxisNotesLabel',
  props: {
    background: String,
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    color: String,
    content: {
      type: Function,
      default: undefined
    },
    font: String,
    format: String,
    position: {
      type: String,
      default: undefined,
      validator: function validator(value) {
        return ['inside', 'outside'].includes(value);
      }
    },
    rotation: Number,
    visible: {
      type: Boolean,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'label'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'label'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartValueAxisNotesLabel = ChartValueAxisNotesLabelVue2;
export { ChartValueAxisNotesLabel, ChartValueAxisNotesLabelVue2 };
