// @progress/kendo-dateinputs-vue-wrapper/dist/es/KendoDatePicker/KendoDatePickerMixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _kendoBaseComponentsVueWrapper = require('@progress/kendo-base-components-vue-wrapper');

var _DatePickerProps = require('./DatePickerProps');

var _DatePickerProps2 = _interopRequireDefault(_DatePickerProps);

var _Animation = require('../Common/Animation');

var _Animation2 = _interopRequireDefault(_Animation);

var _Month = require('../Common/Month');

var _Month2 = _interopRequireDefault(_Month);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    emits: {
        'update:modelValue': null,
        'kendowidgetready': null
    },
    props: _DatePickerProps2.default,
    mixins: [_kendoBaseComponentsVueWrapper.KendoBaseComponent, _kendoBaseComponentsVueWrapper.KendoBaseInputComponent, _Animation2.default, _Month2.default],
    mounted: function mounted() {
        var that = this;
        kendo.jQuery(that.$el).kendoDatePicker(that.widgetOptions);
        var widget = that.kendoWidget();

        if (this.v3) {
            this.ready();
            this.$emit('kendowidgetready', this.kendoWidget());
        } else {
            this.$emit('kendowidgetready', this.kendoWidget());
        }

        that.kendoWidget().bind('change', function () {
            if (that.v3) {
                that.$emit('update:modelValue', widget.value());
            } else {
                that.$emit('changemodel', widget.value());
            }
        });
    },

    methods: {
        kendoWidget: function kendoWidget() {
            return kendo.jQuery(this.$el).getKendoDatePicker();
        }
    }
};
