// @progress/kendo-vue-charts/dist/esm/components/series-item/Outliers.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../base/ConfigurationComponent.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../utils/main.js';
/**
 * @hidden
 */
var ChartSeriesItemOutliersVue2 = {
  name: 'KendoChartSeriesItemOutliers',
  props: {
    background: String,
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    rotation: Number,
    size: Number,
    type: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['square', 'circle', 'triangle', 'cross'].includes(value);
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'outliers'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'outliers'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartSeriesItemOutliers = ChartSeriesItemOutliersVue2;
export { ChartSeriesItemOutliers, ChartSeriesItemOutliersVue2 };
