// @progress/kendo-vue-charts/dist/npm/components/category-axis-item/RangeLabels.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartCategoryAxisRangeLabelsVue2 = exports.ChartCategoryAxisRangeLabels = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./../base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../../utils/main");
/**
 * @hidden
 */
var ChartCategoryAxisRangeLabelsVue2 = {
  name: 'KendoChartCategoryAxisRangeLabels',
  props: {
    position: {
      type: String,
      default: undefined,
      validator: function validator(value) {
        return ['start', 'end', 'onAxis', ''].includes(value);
      }
    },
    skip: Number,
    step: Number,
    visible: {
      type: Boolean,
      default: true
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: config,
      chartKey: 'rangeLabels'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: config,
        chartKey: 'rangeLabels'
      }
    }, [defaultSlots]);
  }
};
exports.ChartCategoryAxisRangeLabelsVue2 = ChartCategoryAxisRangeLabelsVue2;
/**
 * @hidden
 */
var ChartCategoryAxisRangeLabels = ChartCategoryAxisRangeLabelsVue2;
exports.ChartCategoryAxisRangeLabels = ChartCategoryAxisRangeLabels;
