// @progress/kendo-vue-charts/dist/es/components/navigator/Series.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { CollectionConfiguration } from './../base/CollectionConfigurationComponent';
import { getDefaultSlots } from '@progress/kendo-vue-common';
/**
 * @hidden
 */
var ChartNavigatorSeriesVue2 = {
  name: 'KendoChartNavigatorSeries',
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    return (
      // @ts-ignore  function children
      h(CollectionConfiguration, {
        chartKey: "series",
        attrs: this.v3 ? undefined : {
          chartKey: "series",
          parentKey: "navigator"
        },
        parentKey: "navigator"
      }, this.v3 ? function () {
        return [defaultSlots];
      } : [defaultSlots])
    );
  }
};
/**
 * @hidden
 */
var ChartNavigatorSeries = ChartNavigatorSeriesVue2;
export { ChartNavigatorSeries, ChartNavigatorSeriesVue2 };
