// @progress/kendo-dateinputs-vue-wrapper/dist/npm/KendoTimePicker/KendoTimePickerMixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _kendoBaseComponentsVueWrapper = require('@progress/kendo-base-components-vue-wrapper');

var _TimePickerProps = require('./TimePickerProps');

var _TimePickerProps2 = _interopRequireDefault(_TimePickerProps);

var _Animation = require('../Common/Animation');

var _Animation2 = _interopRequireDefault(_Animation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    props: _TimePickerProps2.default,
    emits: {
        'update:modelValue': null,
        'kendowidgetready': null
    },
    mixins: [_kendoBaseComponentsVueWrapper.KendoBaseComponent, _kendoBaseComponentsVueWrapper.KendoBaseInputComponent, _Animation2.default],
    mounted() {
        var that = this;
        kendo.jQuery(that.$el).kendoTimePicker(that.widgetOptions);
        var widget = that.kendoWidget();

        if (this.v3) {
            this.ready();
            this.$emit('kendowidgetready', widget);
        } else {
            this.$emit('kendowidgetready', widget);
        }

        that.kendoWidget().bind('change', function () {
            if (that.v3) {
                that.$emit('update:modelValue', widget.value());
            } else {
                that.$emit('changemodel', widget.value());
            }
        });
    },
    methods: {
        kendoWidget() {
            return kendo.jQuery(this.$el).getKendoTimePicker();
        }
    }
};
