// @progress/kendo-vue-charts/dist/es/components/navigator/category-axis/Labels.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../../base/ConfigurationComponent';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../../utils/main';
/**
 * @hidden
 */
var ChartNavigatorCategoryAxisLabelsVue2 = {
  name: 'KendoChartNavigatorCategoryAxisLabels',
  props: {
    background: String,
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    color: String,
    content: {
      type: Function,
      default: undefined
    },
    culture: String,
    dateFormats: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    font: String,
    format: String,
    margin: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    mirror: {
      type: Boolean,
      default: undefined
    },
    padding: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    position: {
      type: String,
      default: undefined,
      validator: function validator(value) {
        return ['circle', 'column', 'center', 'right', 'left'].includes(value);
      }
    },
    rotation: {
      type: [Object, Number, String],
      default: function _default() {
        return undefined;
      }
    },
    skip: Number,
    step: Number,
    visible: {
      type: Boolean,
      default: true
    },
    visual: {
      type: Function,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'labels'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'labels'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartNavigatorCategoryAxisLabels = ChartNavigatorCategoryAxisLabelsVue2;
export { ChartNavigatorCategoryAxisLabels, ChartNavigatorCategoryAxisLabelsVue2 };
