// @progress/kendo-vue-charts/dist/npm/components/navigator/category-axis/Select.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartNavigatorCategoryAxisSelectVue2 = exports.ChartNavigatorCategoryAxisSelect = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./../../base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../../../utils/main");
/**
 * @hidden
 */
var ChartNavigatorCategoryAxisSelectVue2 = {
  name: 'KendoChartNavigatorCategoryAxisSelect',
  props: {
    from: {
      type: [Object, Number, String],
      default: function _default() {
        return undefined;
      }
    },
    max: {
      type: [Object, Number, String, Date],
      default: function _default() {
        return undefined;
      }
    },
    min: {
      type: [Object, Number, String, Date],
      default: function _default() {
        return undefined;
      }
    },
    mousewheel: {
      type: [Object, Boolean],
      default: function _default() {
        return undefined;
      }
    },
    to: {
      type: [Object, Number, String],
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: config,
      chartKey: 'select'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: config,
        chartKey: 'select'
      }
    }, [defaultSlots]);
  }
};
exports.ChartNavigatorCategoryAxisSelectVue2 = ChartNavigatorCategoryAxisSelectVue2;
/**
 * @hidden
 */
var ChartNavigatorCategoryAxisSelect = ChartNavigatorCategoryAxisSelectVue2;
exports.ChartNavigatorCategoryAxisSelect = ChartNavigatorCategoryAxisSelect;
