// @progress/kendo-vue-charts/dist/npm/Chart.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartVue2 = exports.Chart = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ref = allVue.ref;
var BaseChart_1 = require("./BaseChart");
var DonutCenter_1 = require("./DonutCenter");
var kendo_charts_1 = require("@progress/kendo-charts");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var ChartVue2 = {
  name: 'KendoChart',
  props: {
    donutCenterRender: [String, Function, Object],
    dir: String,
    renderAs: {
      type: String,
      default: undefined,
      validator: function validator(value) {
        return ['svg', 'canvas'].includes(value);
      }
    },
    pannable: {
      type: [Boolean, Object],
      default: function _default() {
        return undefined;
      }
    },
    zoomable: {
      type: [Boolean, Object],
      default: function _default() {
        return undefined;
      }
    },
    seriesColors: {
      type: Array,
      default: function _default() {
        return undefined;
      }
    },
    transitions: {
      type: Boolean,
      default: true
    },
    paneDefaults: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    panes: {
      type: Array,
      default: function _default() {
        return undefined;
      }
    },
    seriesDefaults: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    axisDefaults: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    onAxislabelclick: Function,
    onDrag: Function,
    onDragend: Function,
    onDragstart: Function,
    onLegenditemhover: Function,
    onLegenditemclick: Function,
    onNoteclick: Function,
    onNotehover: Function,
    onPlotareaclick: Function,
    onPlotareahover: Function,
    onRender: Function,
    onSelect: Function,
    onSelectend: Function,
    onSelectstart: Function,
    onSeriesclick: Function,
    onSerieshover: Function,
    onZoom: Function,
    onZoomend: Function,
    onZoomstart: Function
  },
  updated: function updated() {
    this.baseChart = (0, kendo_vue_common_1.getRef)(this, 'baseChart');
  },
  mounted: function mounted() {
    this.baseChart = (0, kendo_vue_common_1.getRef)(this, 'baseChart');
  },
  methods: {
    chartInstance: function chartInstance() {
      if (this.baseChart !== null) {
        return this.baseChart.chartInstance;
      }
      return null;
    },
    surface: function surface() {
      if (this.baseChart !== null) {
        return this.baseChart.surface;
      }
      return null;
    },
    element: function element() {
      if (this.baseChart !== null) {
        return this.baseChart.element;
      }
      return null;
    },
    getTarget: function getTarget() {
      return this;
    },
    handleBaseChartEvents: function handleBaseChartEvents(name, options) {
      this.$emit(name, options);
    },
    handleLegendItemClick: function handleLegendItemClick(event) {
      this.$emit('legenditemclick', event);
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var _this = this;
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var _a = this.$props,
      donutCenterRender = _a.donutCenterRender,
      dir = _a.dir,
      renderAs = _a.renderAs,
      pannable = _a.pannable,
      zoomable = _a.zoomable,
      seriesColors = _a.seriesColors,
      transitions = _a.transitions,
      paneDefaults = _a.paneDefaults,
      panes = _a.panes,
      seriesDefaults = _a.seriesDefaults,
      axisDefaults = _a.axisDefaults;
    var centerRender = donutCenterRender ? kendo_vue_common_1.templateRendering.call(this, donutCenterRender, kendo_vue_common_1.getListeners.call(this)) : null;
    return (
      // @ts-ignore  function children
      h(BaseChart_1.BaseChart, {
        dir: dir,
        attrs: this.v3 ? undefined : {
          dir: dir,
          renderAs: renderAs,
          pannable: pannable,
          zoomable: zoomable,
          seriesColors: seriesColors,
          transitions: transitions,
          paneDefaults: paneDefaults,
          panes: panes,
          seriesDefaults: seriesDefaults,
          axisDefaults: axisDefaults,
          chartConstructor: kendo_charts_1.Chart,
          getTarget: this.getTarget,
          wrapper: "div",
          allListeners: kendo_vue_common_1.getListeners.call(this)
        },
        renderAs: renderAs,
        pannable: pannable,
        zoomable: zoomable,
        seriesColors: seriesColors,
        transitions: transitions,
        paneDefaults: paneDefaults,
        panes: panes,
        seriesDefaults: seriesDefaults,
        axisDefaults: axisDefaults,
        ref: (0, kendo_vue_common_1.setRef)(this, 'baseChart'),
        chartConstructor: kendo_charts_1.Chart,
        getTarget: this.getTarget,
        wrapper: "div",
        "class": (0, kendo_vue_common_1.classNames)('k-chart k-widget'),
        onLegenditemclick: this.handleLegendItemClick,
        on: this.v3 ? undefined : {
          "legenditemclick": this.handleLegendItemClick,
          "basechartevent": this.handleBaseChartEvents
        },
        onBasechartevent: this.handleBaseChartEvents,
        allListeners: kendo_vue_common_1.getListeners.call(this)
      }, this.v3 ? function () {
        return [defaultSlots,
        // @ts-ignore
        h(DonutCenter_1.DonutCenter, {
          render: centerRender,
          attrs: _this.v3 ? undefined : {
            render: centerRender
          }
        })];
      } : [defaultSlots, h(DonutCenter_1.DonutCenter, {
        render: centerRender,
        attrs: _this.v3 ? undefined : {
          render: centerRender
        }
      })])
    );
  }
};
exports.ChartVue2 = ChartVue2;
/**
 * @hidden
 */
var Chart = ChartVue2;
exports.Chart = Chart;
