// @progress/kendo-vue-charts/dist/npm/Sparkline.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SparklineVue2 = exports.Sparkline = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ref = allVue.ref;
var BaseChart_1 = require("./BaseChart");
var kendo_charts_1 = require("@progress/kendo-charts");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var SparklineVue2 = {
  name: 'KendoSparkline',
  props: {
    dataItems: {
      type: Array,
      default: function _default() {
        return undefined;
      }
    },
    type: String,
    dir: String,
    renderAs: {
      type: String,
      default: undefined,
      validator: function validator(value) {
        return ['svg', 'canvas'].includes(value);
      }
    },
    pannable: {
      type: [Boolean, Object],
      default: function _default() {
        return undefined;
      }
    },
    zoomable: {
      type: [Boolean, Object],
      default: function _default() {
        return undefined;
      }
    },
    seriesColors: {
      type: Array,
      default: function _default() {
        return undefined;
      }
    },
    transitions: {
      type: Boolean,
      default: true
    },
    paneDefaults: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    panes: {
      type: Array,
      default: function _default() {
        return undefined;
      }
    },
    seriesDefaults: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    axisDefaults: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    }
  },
  updated: function updated() {
    this.baseChart = (0, kendo_vue_common_1.getRef)(this, 'baseChart');
  },
  mounted: function mounted() {
    this.baseChart = (0, kendo_vue_common_1.getRef)(this, 'baseChart');
  },
  methods: {
    chartInstance: function chartInstance() {
      if (this.baseChart !== null) {
        return this.baseChart.chartInstance;
      }
      return null;
    },
    surface: function surface() {
      if (this.baseChart !== null) {
        return this.baseChart.surface;
      }
      return null;
    },
    element: function element() {
      if (this.baseChart !== null) {
        return this.baseChart.element;
      }
      return null;
    },
    getTarget: function getTarget() {
      return this;
    },
    deriveOptionsFromParent: function deriveOptionsFromParent(options) {
      var _a = this.$props,
        type = _a.type,
        dataItems = _a.dataItems;
      var sparklineOptions = Object.assign({}, options, {
        type: type,
        data: dataItems
      });
      return kendo_charts_1.Sparkline.normalizeOptions(sparklineOptions);
    },
    handleBaseChartEvents: function handleBaseChartEvents(name, options) {
      this.$emit(name, options);
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var _a = this.$props,
      dataItems = _a.dataItems,
      dir = _a.dir,
      renderAs = _a.renderAs,
      pannable = _a.pannable,
      zoomable = _a.zoomable,
      seriesColors = _a.seriesColors,
      transitions = _a.transitions,
      paneDefaults = _a.paneDefaults,
      panes = _a.panes,
      seriesDefaults = _a.seriesDefaults,
      axisDefaults = _a.axisDefaults;
    return (
      // @ts-ignore  function children
      h(BaseChart_1.BaseChart, {
        dataItems: dataItems,
        attrs: this.v3 ? undefined : {
          dataItems: dataItems,
          dir: dir,
          renderAs: renderAs,
          pannable: pannable,
          zoomable: zoomable,
          seriesColors: seriesColors,
          transitions: transitions,
          paneDefaults: paneDefaults,
          panes: panes,
          seriesDefaults: seriesDefaults,
          axisDefaults: axisDefaults,
          chartConstructor: kendo_charts_1.Sparkline,
          getTarget: this.getTarget,
          wrapper: "span",
          deriveOptionsFromParent: this.deriveOptionsFromParent
        },
        dir: dir,
        renderAs: renderAs,
        pannable: pannable,
        zoomable: zoomable,
        seriesColors: seriesColors,
        transitions: transitions,
        paneDefaults: paneDefaults,
        panes: panes,
        seriesDefaults: seriesDefaults,
        axisDefaults: axisDefaults,
        ref: (0, kendo_vue_common_1.setRef)(this, 'baseChart'),
        chartConstructor: kendo_charts_1.Sparkline,
        getTarget: this.getTarget,
        wrapper: "span",
        deriveOptionsFromParent: this.deriveOptionsFromParent,
        "class": (0, kendo_vue_common_1.classNames)('k-sparkline k-widget'),
        onBasechartevent: this.handleBaseChartEvents,
        on: this.v3 ? undefined : {
          "basechartevent": this.handleBaseChartEvents
        }
      }, this.v3 ? function () {
        return [defaultSlots];
      } : [defaultSlots])
    );
  }
};
exports.SparklineVue2 = SparklineVue2;
/**
 * @hidden
 */
var Sparkline = SparklineVue2;
exports.Sparkline = Sparkline;
