// @progress/kendo-vue-animation/dist/esm/Fade.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Animation } from './Animation.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
/**
 * @hidden
 */
var FadeVue2 = {
  props: {
    appear: {
      type: Boolean,
      default: false
    },
    enter: {
      type: Boolean,
      default: true
    },
    exit: {
      type: Boolean,
      default: false
    },
    transitionEnterDuration: {
      type: Number,
      default: 500
    },
    transitionExitDuration: {
      type: Number,
      default: 500
    },
    componentChildStyle: Object,
    childFactory: Object,
    className: String,
    tag: String,
    id: String
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = getDefaultSlots(this);
    return (
      // @ts-ignore function children
      h(Animation, {
        transitionName: "fade",
        attrs: this.v3 ? undefined : {
          transitionName: "fade",
          appear: this.$props.appear,
          enter: this.$props.enter,
          exit: this.$props.exit,
          componentChildStyle: this.$props.componentChildStyle,
          transitionEnterDuration: this.$props.transitionEnterDuration,
          transitionExitDuration: this.$props.transitionExitDuration
        },
        appear: this.$props.appear,
        enter: this.$props.enter,
        exit: this.$props.exit,
        componentChildStyle: this.$props.componentChildStyle,
        transitionEnterDuration: this.$props.transitionEnterDuration,
        transitionExitDuration: this.$props.transitionExitDuration
      }, this.v3 ? function () {
        return [defaultSlot];
      } : [defaultSlot])
    );
  }
};
/**
 * @hidden
 */
var Fade = FadeVue2;
export { Fade, FadeVue2 };
