// @progress/kendo-vue-charts/dist/es/components/axis-defaults/Crosshair.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../base/ConfigurationComponent';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../utils/main';
/**
 * @hidden
 */
var ChartAxisDefaultsCrosshairVue2 = {
  name: 'KendoChartAxisDefaultsCrosshair',
  props: {
    color: String,
    dashType: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['dash', 'dashDot', 'dot', 'longDash', 'longDashDot', 'longDashDotDot', 'solid'].includes(value);
      }
    },
    opacity: Number,
    visible: {
      type: Boolean,
      default: true
    },
    width: Number,
    tooltip: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'crosshair'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'crosshair'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartAxisDefaultsCrosshair = ChartAxisDefaultsCrosshairVue2;
export { ChartAxisDefaultsCrosshair, ChartAxisDefaultsCrosshairVue2 };
