// @progress/kendo-vue-charts/dist/es/components/legend/Item.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../base/ConfigurationComponent';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../utils/main';
/**
 * @hidden
 */
var ChartLegendItemVue2 = {
  name: 'KendoChartLegendItem',
  props: {
    cursor: String,
    type: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['area', 'line'].includes(value);
      }
    },
    line: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    area: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    markers: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    highlight: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    visual: {
      type: Function,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'item'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'item'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartLegendItem = ChartLegendItemVue2;
export { ChartLegendItem, ChartLegendItemVue2 };
