// @progress/kendo-vue-charts/dist/es/components/category-axis-item/CrosshairTooltip.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../base/ConfigurationComponent';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../utils/main';
/**
 * @hidden
 */
var ChartCategoryAxisCrosshairTooltipVue2 = {
  name: 'KendoChartCategoryAxisCrosshairTooltip',
  props: {
    background: String,
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    color: String,
    font: String,
    format: String,
    padding: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    visible: {
      type: Boolean,
      default: true
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'tooltip'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'tooltip'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartCategoryAxisCrosshairTooltip = ChartCategoryAxisCrosshairTooltipVue2;
export { ChartCategoryAxisCrosshairTooltip, ChartCategoryAxisCrosshairTooltipVue2 };
