// @progress/kendo-vue-charts/dist/es/components/category-axis-item/RangeLabels.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../base/ConfigurationComponent';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../utils/main';
/**
 * @hidden
 */
var ChartCategoryAxisRangeLabelsVue2 = {
  name: 'KendoChartCategoryAxisRangeLabels',
  props: {
    position: {
      type: String,
      default: undefined,
      validator: function validator(value) {
        return ['start', 'end', 'onAxis', ''].includes(value);
      }
    },
    skip: Number,
    step: Number,
    visible: {
      type: Boolean,
      default: true
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'rangeLabels'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'rangeLabels'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartCategoryAxisRangeLabels = ChartCategoryAxisRangeLabelsVue2;
export { ChartCategoryAxisRangeLabels, ChartCategoryAxisRangeLabelsVue2 };
