// @progress/kendo-vue-intl/dist/esm/Localization/LocalizationProvider.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ref = allVue.ref;
var provide = allVue.provide;
import { LocalizationService } from './LocalizationService.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
/**
 * @hidden
 */
var LocalizationProviderVue2 = {
  props: {
    language: String
  },
  data: function data() {
    return this.getChildContext();
  },
  watch: {
    language: function language(newLanguage) {
      this.$data.kendoLocalizationService.language = newLanguage;
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function (props) {
    var v3 = !!isV3;
    var localizationService = ref(new LocalizationService(props.language));
    provide('kendoLocalizationService', localizationService);
    return {
      v3: v3
    };
  },
  provide: function provide() {
    return {
      kendoLocalizationService: this.$data.kendoLocalizationService
    };
  },
  methods: {
    /**
     * Returns a localization service.
     * The method is suitable for overriding when you
     * implement custom localization behavior.
     */
    getLocalizationService: function getLocalizationService() {
      return new LocalizationService(this.$props.language);
    },
    /**
     * @hidden
     */
    getChildContext: function getChildContext() {
      return {
        kendoLocalizationService: this.getLocalizationService()
      };
    }
  },
  /**
   * @hidden
   */
  // @ts-ignore
  render: function render(createElement) {
    // @ts-ignore
    var h = gh || createElement;
    var defaultSlot = getDefaultSlots(this);
    return h("div", [defaultSlot]);
  }
};
/**
 *
 *
 * A Vue component which provides a localization service. Expects a language string as a property of the component.
 *
 *
 * ### props
 * The props of the LocalizationProvider component.
 *
 *
 *
 * ## Methods
 *
 * ### getLocalizationService
 * Returns a localization service. The method is suitable for overriding when you implement custom localization behavior.
 *
 * #### Returns
 * <span class='code'>[LocalizationService]({% slug api_intl_localizationservice %})</span>
 */
var LocalizationProvider = LocalizationProviderVue2;
export { LocalizationProvider, LocalizationProviderVue2 };
