// @progress/kendo-vue-charts/dist/npm/components/series-item/Labels.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartSeriesLabelsVue2 = exports.ChartSeriesLabels = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./../base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../../utils/main");
/**
 * @hidden
 */
var ChartSeriesLabelsVue2 = {
  name: 'KendoChartSeriesLabels',
  props: {
    align: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['circle', 'column', 'center', 'right', 'left'].includes(value);
      }
    },
    background: String,
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    color: String,
    content: {
      type: Function,
      default: undefined
    },
    distance: Number,
    font: String,
    format: String,
    margin: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    padding: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    position: {
      type: String,
      default: undefined,
      validator: function validator(value) {
        return ['auto', 'above', 'below', 'center', 'insideBase', 'insideEnd', 'left', 'outsideEnd', 'right', 'top', 'bottom'].includes(value);
      }
    },
    visible: {
      type: Boolean,
      default: true
    },
    visual: {
      type: Function,
      default: undefined
    },
    from: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    to: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: config,
      chartKey: 'labels'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: config,
        chartKey: 'labels'
      }
    }, [defaultSlots]);
  }
};
exports.ChartSeriesLabelsVue2 = ChartSeriesLabelsVue2;
/**
 * @hidden
 */
var ChartSeriesLabels = ChartSeriesLabelsVue2;
exports.ChartSeriesLabels = ChartSeriesLabels;
