// @progress/kendo-vue-charts/dist/npm/components/series-item/NotesLabel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartSeriesNotesLabelVue2 = exports.ChartSeriesNotesLabel = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./../base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../../utils/main");
/**
 * @hidden
 */
var ChartSeriesNotesLabelVue2 = {
  name: 'KendoChartSeriesNotesLabel',
  props: {
    background: String,
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    color: String,
    content: {
      type: Function,
      default: undefined
    },
    font: String,
    format: String,
    position: {
      type: String,
      default: undefined,
      validator: function validator(value) {
        return ['inside', 'outside'].includes(value);
      }
    },
    rotation: Number,
    visible: {
      type: Boolean,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: config,
      chartKey: 'label'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: config,
        chartKey: 'label'
      }
    }, [defaultSlots]);
  }
};
exports.ChartSeriesNotesLabelVue2 = ChartSeriesNotesLabelVue2;
/**
 * @hidden
 */
var ChartSeriesNotesLabel = ChartSeriesNotesLabelVue2;
exports.ChartSeriesNotesLabel = ChartSeriesNotesLabel;
