// @progress/kendo-vue-charts/dist/npm/components/y-axis-item/Crosshair.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartYAxisCrosshairVue2 = exports.ChartYAxisCrosshair = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./../base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../../utils/main");
/**
 * @hidden
 */
var ChartYAxisCrosshairVue2 = {
  name: 'KendoChartYAxisCrosshair',
  props: {
    color: String,
    opacity: Number,
    visible: {
      type: Boolean,
      default: true
    },
    width: Number,
    tooltip: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: config,
      chartKey: 'crosshair'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: config,
        chartKey: 'crosshair'
      }
    }, [defaultSlots]);
  }
};
exports.ChartYAxisCrosshairVue2 = ChartYAxisCrosshairVue2;
/**
 * @hidden
 */
var ChartYAxisCrosshair = ChartYAxisCrosshairVue2;
exports.ChartYAxisCrosshair = ChartYAxisCrosshair;
