// @progress/kendo-vue-charts/dist/npm/components/y-axis-item/Labels.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartYAxisLabelsVue2 = exports.ChartYAxisLabels = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./../base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../../utils/main");
/**
 * @hidden
 */
var ChartYAxisLabelsVue2 = {
  name: 'KendoChartYAxisLabels',
  props: {
    background: String,
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    color: String,
    content: {
      type: Function,
      default: undefined
    },
    culture: String,
    dateFormats: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    font: String,
    format: String,
    margin: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    mirror: {
      type: Boolean,
      default: undefined
    },
    padding: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    rotation: {
      type: [Object, Number, String],
      default: function _default() {
        return undefined;
      }
    },
    skip: Number,
    step: Number,
    visible: {
      type: Boolean,
      default: true
    },
    visual: {
      type: Function,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: config,
      chartKey: 'labels'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: config,
        chartKey: 'labels'
      }
    }, [defaultSlots]);
  }
};
exports.ChartYAxisLabelsVue2 = ChartYAxisLabelsVue2;
/**
 * @hidden
 */
var ChartYAxisLabels = ChartYAxisLabelsVue2;
exports.ChartYAxisLabels = ChartYAxisLabels;
