// @progress/kendo-vue-charts/dist/esm/components/pane/Title.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../base/ConfigurationComponent.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../utils/main.js';
/**
 * @hidden
 */
var ChartPaneTitleVue2 = {
  name: 'KendoChartPaneTitle',
  props: {
    background: String,
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    color: String,
    font: String,
    margin: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    position: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['center', 'left', 'right'].includes(value);
      }
    },
    text: String,
    visible: {
      type: Boolean,
      default: undefined
    },
    visual: {
      type: Function,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'title'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'title'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartPaneTitle = ChartPaneTitleVue2;
export { ChartPaneTitle, ChartPaneTitleVue2 };
