// @progress/kendo-vue-charts/dist/npm/components/navigator/series-item/Outliers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartNavigatorSeriesItemOutliersVue2 = exports.ChartNavigatorSeriesItemOutliers = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./../../base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../../../utils/main");
/**
 * @hidden
 */
var ChartNavigatorSeriesItemOutliersVue2 = {
  name: 'KendoChartNavigatorSeriesItemOutliers',
  props: {
    background: String,
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    rotation: Number,
    size: Number,
    type: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['square', 'circle', 'triangle', 'cross'].includes(value);
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: config,
      chartKey: 'outliers'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: config,
        chartKey: 'outliers'
      }
    }, [defaultSlots]);
  }
};
exports.ChartNavigatorSeriesItemOutliersVue2 = ChartNavigatorSeriesItemOutliersVue2;
/**
 * @hidden
 */
var ChartNavigatorSeriesItemOutliers = ChartNavigatorSeriesItemOutliersVue2;
exports.ChartNavigatorSeriesItemOutliers = ChartNavigatorSeriesItemOutliers;
