// @progress/kendo-vue-charts/dist/npm/components/navigator/series-item/Tooltip.js
"use strict";

var __assign = undefined && undefined.__assign || function () {
  __assign = Object.assign || function (t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];
      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
      }
    }
    return t;
  };
  return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartNavigatorSeriesItemTooltipVue2 = exports.ChartNavigatorSeriesItemTooltip = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./../../base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../../../utils/main");
/**
 * @hidden
 */
var ChartNavigatorSeriesItemTooltipVue2 = {
  name: 'KendoChartNavigatorSeriesItemTooltip',
  props: {
    background: String,
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    color: String,
    font: String,
    format: String,
    padding: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    render: [String, Function, Object],
    visible: {
      type: Boolean,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    var render = kendo_vue_common_1.templateRendering.call(this, this.$props.render, kendo_vue_common_1.getListeners.call(this));
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: __assign(__assign({}, config), {
        render: render
      }),
      chartKey: 'tooltip'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: __assign(__assign({}, config), {
          render: render
        }),
        chartKey: 'tooltip'
      }
    }, [defaultSlots]);
  }
};
exports.ChartNavigatorSeriesItemTooltipVue2 = ChartNavigatorSeriesItemTooltipVue2;
/**
 * @hidden
 */
var ChartNavigatorSeriesItemTooltip = ChartNavigatorSeriesItemTooltipVue2;
exports.ChartNavigatorSeriesItemTooltip = ChartNavigatorSeriesItemTooltip;
