// @progress/kendo-vue-charts/dist/esm/components/navigator/series-item/Highlight.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../../base/ConfigurationComponent.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../../utils/main.js';
/**
 * @hidden
 */
var ChartNavigatorSeriesHighlightVue2 = {
  name: 'KendoChartNavigatorSeriesHighlight',
  props: {
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    color: String,
    line: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    opacity: Number,
    toggle: {
      type: Function,
      default: undefined
    },
    visible: {
      type: Boolean,
      default: undefined
    },
    visual: {
      type: Function,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'highlight'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'highlight'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartNavigatorSeriesHighlight = ChartNavigatorSeriesHighlightVue2;
export { ChartNavigatorSeriesHighlight, ChartNavigatorSeriesHighlightVue2 };
