// @progress/kendo-vue-charts/dist/esm/components/navigator/series-item/Notes.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../../base/ConfigurationComponent.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../../utils/main.js';
/**
 * @hidden
 */
var ChartNavigatorSeriesNotesVue2 = {
  name: 'KendoChartNavigatorSeriesNotes',
  props: {
    line: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    position: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['top', 'bottom', 'left', 'right'].includes(value);
      }
    },
    visual: {
      type: Function,
      default: undefined
    },
    icon: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    label: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'notes'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'notes'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartNavigatorSeriesNotes = ChartNavigatorSeriesNotesVue2;
export { ChartNavigatorSeriesNotes, ChartNavigatorSeriesNotesVue2 };
