// @progress/kendo-vue-animation/dist/npm/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cancelAnimation = exports.animate = exports.RevealVue2 = exports.Reveal = exports.ZoomVue2 = exports.Zoom = exports.SlideVue2 = exports.Slide = exports.PushVue2 = exports.Push = exports.ExpandVue2 = exports.Expand = exports.FadeVue2 = exports.Fade = exports.AnimationChildVue2 = exports.AnimationChild = exports.AnimationVue2 = exports.Animation = void 0;
var Animation_1 = require("./Animation");
Object.defineProperty(exports, "Animation", { enumerable: true, get: function () { return Animation_1.Animation; } });
Object.defineProperty(exports, "AnimationVue2", { enumerable: true, get: function () { return Animation_1.AnimationVue2; } });
var AnimationChild_1 = require("./AnimationChild");
Object.defineProperty(exports, "AnimationChild", { enumerable: true, get: function () { return AnimationChild_1.AnimationChild; } });
Object.defineProperty(exports, "AnimationChildVue2", { enumerable: true, get: function () { return AnimationChild_1.AnimationChildVue2; } });
var Fade_1 = require("./Fade");
Object.defineProperty(exports, "Fade", { enumerable: true, get: function () { return Fade_1.Fade; } });
Object.defineProperty(exports, "FadeVue2", { enumerable: true, get: function () { return Fade_1.FadeVue2; } });
var Expand_1 = require("./Expand");
Object.defineProperty(exports, "Expand", { enumerable: true, get: function () { return Expand_1.Expand; } });
Object.defineProperty(exports, "ExpandVue2", { enumerable: true, get: function () { return Expand_1.ExpandVue2; } });
var Push_1 = require("./Push");
Object.defineProperty(exports, "Push", { enumerable: true, get: function () { return Push_1.Push; } });
Object.defineProperty(exports, "PushVue2", { enumerable: true, get: function () { return Push_1.PushVue2; } });
var Slide_1 = require("./Slide");
Object.defineProperty(exports, "Slide", { enumerable: true, get: function () { return Slide_1.Slide; } });
Object.defineProperty(exports, "SlideVue2", { enumerable: true, get: function () { return Slide_1.SlideVue2; } });
var Zoom_1 = require("./Zoom");
Object.defineProperty(exports, "Zoom", { enumerable: true, get: function () { return Zoom_1.Zoom; } });
Object.defineProperty(exports, "ZoomVue2", { enumerable: true, get: function () { return Zoom_1.ZoomVue2; } });
var Reveal_1 = require("./Reveal");
Object.defineProperty(exports, "Reveal", { enumerable: true, get: function () { return Reveal_1.Reveal; } });
Object.defineProperty(exports, "RevealVue2", { enumerable: true, get: function () { return Reveal_1.RevealVue2; } });
var useAnimation_1 = require("./hooks/useAnimation");
Object.defineProperty(exports, "animate", { enumerable: true, get: function () { return useAnimation_1.animate; } });
Object.defineProperty(exports, "cancelAnimation", { enumerable: true, get: function () { return useAnimation_1.cancelAnimation; } });
