// @progress/kendo-vue-charts/dist/es/components/Zoomable.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './base/ConfigurationComponent';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../utils/main';
/**
 * @hidden
 */
var ChartZoomableVue2 = {
  name: 'KendoChartZoomable',
  props: {
    mousewheel: {
      type: [Boolean, Object],
      default: function _default() {
        return undefined;
      }
    },
    selection: {
      type: [Boolean, Object],
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'zoomable'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'zoomable'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartZoomable = ChartZoomableVue2;
export { ChartZoomable, ChartZoomableVue2 };
