// @progress/kendo-vue-charts/dist/esm/components/navigator/category-axis/Select.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../../base/ConfigurationComponent.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../../utils/main.js';
/**
 * @hidden
 */
var ChartNavigatorCategoryAxisSelectVue2 = {
  name: 'KendoChartNavigatorCategoryAxisSelect',
  props: {
    from: {
      type: [Object, Number, String],
      default: function _default() {
        return undefined;
      }
    },
    max: {
      type: [Object, Number, String, Date],
      default: function _default() {
        return undefined;
      }
    },
    min: {
      type: [Object, Number, String, Date],
      default: function _default() {
        return undefined;
      }
    },
    mousewheel: {
      type: [Object, Boolean],
      default: function _default() {
        return undefined;
      }
    },
    to: {
      type: [Object, Number, String],
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'select'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'select'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartNavigatorCategoryAxisSelect = ChartNavigatorCategoryAxisSelectVue2;
export { ChartNavigatorCategoryAxisSelect, ChartNavigatorCategoryAxisSelectVue2 };
