// @progress/kendo-vue-intl/dist/esm/Intl/IntlProvider.js
// @ts-ignore
import * as aVue from 'vue';
var allVue = aVue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ref = allVue.ref;
var provide = allVue.provide;
import { IntlService } from './IntlService.js';
import { getDefaultSlots, validatePackage } from '@progress/kendo-vue-common';
import { packageMetadata } from '../package-metadata.js';
/**
 * @hidden
 */
var IntlProviderVue2 = {
  props: {
    locale: String
  },
  data: function data() {
    return this.getChildContext();
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function (props) {
    var v3 = !!isV3;
    var intlService = ref(new IntlService(props.locale));
    provide('kendoIntlService', intlService);
    return {
      v3: v3
    };
  },
  watch: {
    locale: function locale(newLocale) {
      this.$data.kendoIntlService.locale = newLocale;
    }
  },
  provide: function provide() {
    return {
      kendoIntlService: this.$data.kendoIntlService
    };
  },
  methods: {
    /**
     * Returns an internationalization service.
     * The method is suitable for overriding when you
     * implement custom internationalization behavior.
     */
    getIntlService: function getIntlService() {
      return new IntlService(this.$props.locale);
    },
    /**
     * @hidden
     */
    getChildContext: function getChildContext() {
      return {
        kendoIntlService: this.getIntlService()
      };
    }
  },
  created: function created() {
    validatePackage(packageMetadata);
  },
  /**
   * @hidden
   */
  // @ts-ignore
  render: function render(createElement) {
    // @ts-ignore
    var h = gh || createElement;
    var defaultSlot = getDefaultSlots(this);
    return h("div", [defaultSlot]);
  }
};
/**
 *
 * A Vue component which provides an internationalization service. Expects a locale string as a property of the component.
 *
 *
 * ### props
 * The props of the IntlProvider component.
 *
 *
 *
 * ## Methods
 *
 * ### getIntlService
 * Returns an internationalization service. The method is suitable for overriding when you implement custom internationalization behavior.
 *
 * #### Returns
 * <span class='code'>[IntlService]({% slug api_intl_intlservice %})</span>
 */
var IntlProvider = IntlProviderVue2;
export { IntlProvider, IntlProviderVue2 };
