// @progress/kendo-vue-charts/dist/esm/components/CategoryAxis.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { CollectionConfiguration } from './base/CollectionConfigurationComponent.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
/**
 * @hidden
 */
var ChartCategoryAxisVue2 = {
  name: 'KendoChartCategoryAxis',
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    return (
      // @ts-ignore  function children
      h(CollectionConfiguration, {
        chartKey: "categoryAxis",
        attrs: this.v3 ? undefined : {
          chartKey: "categoryAxis"
        }
      }, this.v3 ? function () {
        return [defaultSlots];
      } : [defaultSlots])
    );
  }
};
/**
 * @hidden
 */
var ChartCategoryAxis = ChartCategoryAxisVue2;
export { ChartCategoryAxis, ChartCategoryAxisVue2 };
