// @progress/kendo-vue-charts/dist/esm/components/ChartArea.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './base/ConfigurationComponent.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../utils/main.js';
/**
 * @hidden
 */
var ChartAreaVue2 = {
  name: 'KendoChartArea',
  props: {
    background: {
      type: String,
      default: undefined
    },
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    height: {
      type: Number,
      default: undefined
    },
    margin: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    opacity: {
      type: Number,
      default: undefined
    },
    width: {
      type: Number,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'chartArea'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'chartArea'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartArea = ChartAreaVue2;
export { ChartArea, ChartAreaVue2 };
