// @progress/kendo-vue-charts/dist/esm/components/Legend.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './base/ConfigurationComponent.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../utils/main.js';
/**
 * @hidden
 */
var ChartLegendVue2 = {
  name: 'KendoChartLegend',
  props: {
    align: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['start', 'center', 'end'].includes(value);
      }
    },
    background: {
      type: String,
      default: undefined
    },
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    height: {
      type: Number,
      default: undefined
    },
    labels: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    margin: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    offsetX: {
      type: Number,
      default: undefined
    },
    offsetY: {
      type: Number,
      default: undefined
    },
    orientation: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['vertical', 'horizontal'].includes(value);
      }
    },
    padding: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    position: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['top', 'bottom', 'left', 'right', 'custom'].includes(value);
      }
    },
    reverse: {
      type: Boolean,
      default: undefined
    },
    visible: {
      type: Boolean,
      default: true
    },
    width: {
      type: Number,
      default: undefined
    },
    inactiveItems: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    item: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    spacing: {
      type: Number,
      default: undefined
    },
    title: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'legend'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'legend'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartLegend = ChartLegendVue2;
export { ChartLegend, ChartLegendVue2 };
