// @progress/kendo-vue-charts/dist/esm/components/Navigator.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './base/ConfigurationComponent.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../utils/main.js';
/**
 * @hidden
 */
var ChartNavigatorVue2 = {
  name: 'KendoChartNavigator',
  props: {
    visible: {
      type: Boolean,
      default: true
    },
    categoryAxis: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    hint: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    pane: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    select: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    series: {
      type: [Object, Array],
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'navigator'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'navigator'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartNavigator = ChartNavigatorVue2;
export { ChartNavigator, ChartNavigatorVue2 };
