// @progress/kendo-vue-charts/dist/esm/components/ValueAxisItem.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './base/ConfigurationComponent.js';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../utils/main.js';
/**
 * @hidden
 */
var ChartValueAxisItemVue2 = {
  name: 'KendoChartValueAxisItem',
  props: {
    axisCrossingValue: {
      type: [Array, Object, Date, Number],
      default: function _default() {
        return undefined;
      }
    },
    background: {
      type: String,
      default: undefined
    },
    color: {
      type: String,
      default: undefined
    },
    line: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    majorGridLines: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    majorTicks: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    majorUnit: {
      type: Number,
      default: undefined
    },
    max: {
      type: [Number, Object, String, Date],
      default: function _default() {
        return undefined;
      }
    },
    min: {
      type: [Number, Object, String, Date],
      default: function _default() {
        return undefined;
      }
    },
    minorGridLines: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    minorTicks: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    minorUnit: {
      type: Number,
      default: undefined
    },
    name: {
      type: String,
      default: undefined
    },
    narrowRange: {
      type: Boolean,
      default: function _default() {
        return undefined;
      }
    },
    pane: {
      type: String,
      default: undefined
    },
    plotBands: {
      type: Array,
      default: function _default() {
        return undefined;
      }
    },
    reverse: {
      type: Boolean,
      default: function _default() {
        return undefined;
      }
    },
    type: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['numeric', 'log'].includes(value);
      }
    },
    visible: {
      type: Boolean,
      default: function _default() {
        return undefined;
      }
    },
    crosshair: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    labels: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    notes: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    title: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartValueAxisItem = ChartValueAxisItemVue2;
export { ChartValueAxisItem, ChartValueAxisItemVue2 };
