// @progress/kendo-vue-charts/dist/npm/components/AxisDefaults.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartAxisDefaultsVue2 = exports.ChartAxisDefaults = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../utils/main");
/**
 * @hidden
 */
var ChartAxisDefaultsVue2 = {
  name: 'KendoChartAxisDefaults',
  props: {
    background: {
      type: String,
      default: undefined
    },
    color: {
      type: String,
      default: undefined
    },
    line: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    majorGridLines: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    majorTicks: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    minorGridLines: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    minorTicks: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    narrowRange: {
      type: Boolean,
      default: undefined
    },
    pane: {
      type: String,
      default: undefined
    },
    plotBands: {
      type: Array,
      default: function _default() {
        return undefined;
      }
    },
    reverse: {
      type: Boolean,
      default: undefined
    },
    startAngle: {
      type: Number,
      default: undefined
    },
    visible: {
      type: Boolean,
      default: undefined
    },
    crosshair: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    labels: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    title: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: config,
      chartKey: 'axisDefaults'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: config,
        chartKey: 'axisDefaults'
      }
    }, [defaultSlots]);
  }
};
exports.ChartAxisDefaultsVue2 = ChartAxisDefaultsVue2;
/**
 * @hidden
 */
var ChartAxisDefaults = ChartAxisDefaultsVue2;
exports.ChartAxisDefaults = ChartAxisDefaults;
