// @progress/kendo-vue-charts/dist/npm/components/Legend.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartLegendVue2 = exports.ChartLegend = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../utils/main");
/**
 * @hidden
 */
var ChartLegendVue2 = {
  name: 'KendoChartLegend',
  props: {
    align: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['start', 'center', 'end'].includes(value);
      }
    },
    background: {
      type: String,
      default: undefined
    },
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    height: {
      type: Number,
      default: undefined
    },
    labels: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    margin: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    offsetX: {
      type: Number,
      default: undefined
    },
    offsetY: {
      type: Number,
      default: undefined
    },
    orientation: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['vertical', 'horizontal'].includes(value);
      }
    },
    padding: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    position: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['top', 'bottom', 'left', 'right', 'custom'].includes(value);
      }
    },
    reverse: {
      type: Boolean,
      default: undefined
    },
    visible: {
      type: Boolean,
      default: true
    },
    width: {
      type: Number,
      default: undefined
    },
    inactiveItems: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    item: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    spacing: {
      type: Number,
      default: undefined
    },
    title: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: config,
      chartKey: 'legend'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: config,
        chartKey: 'legend'
      }
    }, [defaultSlots]);
  }
};
exports.ChartLegendVue2 = ChartLegendVue2;
/**
 * @hidden
 */
var ChartLegend = ChartLegendVue2;
exports.ChartLegend = ChartLegend;
