// @progress/kendo-vue-charts/dist/npm/components/Title.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartTitleVue2 = exports.ChartTitle = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var ConfigurationComponent_1 = require("./base/ConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
var main_1 = require("../utils/main");
/**
 * @hidden
 */
var ChartTitleVue2 = {
  name: 'KendoChartTitle',
  props: {
    background: {
      type: String,
      default: undefined
    },
    border: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    margin: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    padding: {
      type: [Object, Number],
      default: function _default() {
        return undefined;
      }
    },
    align: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['center', 'left', 'right'].includes(value);
      }
    },
    color: {
      type: String,
      default: undefined
    },
    font: {
      type: String,
      default: undefined
    },
    position: {
      type: String,
      default: function _default() {
        return undefined;
      },
      validator: function validator(value) {
        return ['top', 'bottom'].includes(value);
      }
    },
    text: {
      type: String,
      default: undefined
    },
    visible: {
      type: Boolean,
      default: undefined
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    var config = (0, main_1.removeUndefinedProps)(this.$props);
    return this.v3 ? h(ConfigurationComponent_1.Configuration, {
      config: config,
      chartKey: 'title'
    }, function () {
      return defaultSlots;
    }) : h(ConfigurationComponent_1.Configuration, {
      props: {
        config: config,
        chartKey: 'title'
      }
    }, [defaultSlots]);
  }
};
exports.ChartTitleVue2 = ChartTitleVue2;
/**
 * @hidden
 */
var ChartTitle = ChartTitleVue2;
exports.ChartTitle = ChartTitle;
