// @progress/kendo-vue-charts/dist/npm/components/XAxis.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartXAxisVue2 = exports.ChartXAxis = void 0;
// @ts-ignore
var Vue = require("vue");
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
var CollectionConfigurationComponent_1 = require("./base/CollectionConfigurationComponent");
var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */
var ChartXAxisVue2 = {
  name: 'KendoChartXAxis',
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = (0, kendo_vue_common_1.getDefaultSlots)(this);
    return (
      // @ts-ignore  function children
      h(CollectionConfigurationComponent_1.CollectionConfiguration, {
        chartKey: "xAxis",
        attrs: this.v3 ? undefined : {
          chartKey: "xAxis"
        }
      }, this.v3 ? function () {
        return [defaultSlots];
      } : [defaultSlots])
    );
  }
};
exports.ChartXAxisVue2 = ChartXAxisVue2;
/**
 * @hidden
 */
var ChartXAxis = ChartXAxisVue2;
exports.ChartXAxis = ChartXAxis;
