// @progress/kendo-vue-charts/dist/es/components/series-defaults/Notes.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { Configuration } from './../base/ConfigurationComponent';
import { getDefaultSlots } from '@progress/kendo-vue-common';
import { removeUndefinedProps } from '../../utils/main';
/**
 * @hidden
 */
var ChartSeriesDefaultsNotesVue2 = {
  name: 'KendoChartSeriesDefaultsNotes',
  props: {
    line: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    visual: {
      type: Function,
      default: undefined
    },
    icon: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    },
    label: {
      type: Object,
      default: function _default() {
        return undefined;
      }
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = getDefaultSlots(this);
    var config = removeUndefinedProps(this.$props);
    return this.v3 ? h(Configuration, {
      config: config,
      chartKey: 'notes'
    }, function () {
      return defaultSlots;
    }) : h(Configuration, {
      props: {
        config: config,
        chartKey: 'notes'
      }
    }, [defaultSlots]);
  }
};
/**
 * @hidden
 */
var ChartSeriesDefaultsNotes = ChartSeriesDefaultsNotesVue2;
export { ChartSeriesDefaultsNotes, ChartSeriesDefaultsNotesVue2 };
